src/imitation/density_baselines.py

```diff
@@ -0,0 +1,267 @@
+        `density_type=STATE_DENSITY`, since it encourages the agent to imitate
+        entire trajectories instead of simply occupying a few states that have
+        high frequency in the demonstration dataset.
+      standardise_inputs (bool): if True, then the inputs to the reward model
+        will be standardised to have zero mean and unit variance over the
+        demonstration trajectories. Otherwise, inputs will be passed to the
+        reward model with their ordinary scale.
+      kernel (str): kernel to use for density estimation. Current valid options
+        include `gaussian` and `exponential`.
+      kernel_bandwidth (float): bandwidth of kernel. If `standardise_inputs` is
+        true and you are using a Gaussian kernel, then it probably makes sense
+        to set this somewhere between 0.1 and 1."""
+    self.density_type = density_type
+    self.is_stationary = is_stationary
+    self.kernel = kernel
+    self.kernel_bandwidth = kernel_bandwidth
+    self.standardise = standardise_inputs
+    self.obs_space = obs_space
+    self.act_space = act_space
+    self._fit_models(trajectories)
+
+  def _fit_models(self, trajectories):
+    flat_trajs = self._flatten_trajectories(trajectories)
+
+    # if requested, we'll scale demonstration transitions so that they have
+    # zero mean and unit variance (i.e all components are equally important)
+    flattened_dataset = np.stack(sum(flat_trajs, []), axis=0)
+    self._scaler = StandardScaler(with_mean=self.standardise,
+                                  with_std=self.standardise)
+    self._scaler.fit(flattened_dataset)
+
+    # now fit density model
+    # TODO: add absorbing state fix that I describe in __call__
+    if self.is_stationary:
+      # fit to all pairs, since density model is stationary
+      self._density_model = self._fit_single_density(
+          self._scaler.transform(flattened_dataset))
+    else:
+      # fit separately for samples at each time step
+      T = max(map(len, flat_trajs))
+      traj_groups = [[]] * T
+      for traj in flat_trajs:
+        for t, flat_trans in enumerate(traj):
+          traj_groups[t].append(flat_trans)
+      traj_groups_scaled = [
+          self._scaler.transform(np.stack(step_transitions, axis=0))
+          for step_transitions in traj_groups
+      ]
+      self._density_models = [
+          self._fit_single_density(scaled_flat_trans)
+          for scaled_flat_trans in traj_groups_scaled
+      ]
+
+  def _fit_single_density(self, flat_transitions):
+    # This bandwidth was chosen to make sense with standardised inputs that
+    # have unit variance in each component. There might be a better way to
+    # choose it automatically.
+    density_model = KernelDensity(kernel=self.kernel,
+                                  bandwidth=self.kernel_bandwidth)
+    density_model.fit(flat_transitions)
+    return density_model
+
+  def _flatten_trajectories(self, trajectories):
```

Comment:
Doing more than just flattening here. (Totally different behavior from `rollout.flatten_trajectories`). Consider renaming to `process_trajectories`? A brief docstring describing the returned value (params not necessary) would be great too.